drgn_tools/kernfs_memcg.py

```diff
@@ -0,0 +1,224 @@
+
+
+def get_num_active_mem_cgroups(prog: Program) -> int:
+    """
+    Get number of active mem cgroups.
+    """
+    mem_cgroup_subsys = prog["cgroup_subsys"][_MEMORY_CGRP_ID]
+    # add 1 to number of active memcgroups to account for root memcgroup
+    return mem_cgroup_subsys.root.cgrp.nr_descendants.value_() + 1
+
+
+def get_num_dying_mem_cgroups(prog: Program) -> int:
+    """
+    Get number of inactive or dying mem cgroups.
+    """
+    mem_cgroup_subsys = prog["cgroup_subsys"][_MEMORY_CGRP_ID]
+    return mem_cgroup_subsys.root.cgrp.nr_dying_descendants.value_()
+
+
+# By default we scan all pages, that have memcg ref
+# but if max_pages is specified then we bail out
+# after getting those many pages or scanning all
+# pages , whichever happens first
+def dump_page_cache_pages_pinning_cgroups(
+    prog: Program, max_pages: int = 10000
+):
+    """
+    Dump page-cache pages that have reference to a mem-cgroup.
+    The ouput also contains information such as the cgroup that is pinned, its flags
+    (to indicate current state of cgroup) and file cached by this page.
+
+    :params: max_pages: specify how many pages to find. By default first 10000 such
+    pages are listed. Use 0 to list all such pages.
+
+    """
+    PG_slab_mask = 1 << prog.constant("PG_slab")
+    mem_cgroup_root = prog["cgroup_subsys"][_MEMORY_CGRP_ID].root
+    total_count = 0
+    found_count = 0
+    for page in for_each_page(prog):
+        total_count = total_count + 1
+        try:
+            # Ignore slab pages
+            if page.flags & PG_slab_mask:
+                continue
+            # Ignore non page-cache pages
+            if not page.mapping:
+                continue
+            try:
+                mem_cgroup = page.mem_cgroup
+            except AttributeError:
+                mem_cgroup = page.memcg_data
+
+            if not mem_cgroup.value_() or mem_cgroup.value_() & 3:
+                continue
+            cgroup_subsys_state = cast(
+                "struct cgroup_subsys_state *", mem_cgroup
+            )
+            if cgroup_subsys_state.cgroup.root == mem_cgroup_root:
+                found_count = found_count + 1
+                cgrp = cgroup_subsys_state.cgroup
+                address_space = page.mapping
+                inode = address_space.host
+                if inode_path(inode) is None:
+                    continue
+                dentry = container_of(
+                    inode.i_dentry.first, "struct dentry", "d_u.d_alias"
+                )
+                path = dentry_path_any_mount(dentry).decode()
+                cgroup_state = decode_css_flags(cgrp.self.address_of_())
+                print(
+                    f"page: 0x{page.value_():x} cgroup: {cgroup_path(cgrp).decode()} state: {cgroup_state} path: {path}\n"
+                )
+                if max_pages and found_count == max_pages:
+                    break
+        except FaultError:
+            continue
```

Comment:
I don't like silently swallowing errors, but I understand it can be necessary. Would you be willing to at least count the number of occurrences and log it at the end, so that we can understand whether there are actually any issues when the script runs?